File: src/main/java/net/cjsah/console/plugin/PluginLoader.java
package net.cjsah.console.plugin;

import com.google.common.collect.Maps;
import com.google.gson.JsonObject;
import net.cjsah.console.Console;
import net.cjsah.console.ConsoleFiles;
import net.cjsah.console.Util;
import net.cjsah.console.exceptions.PluginException;

import java.io.File;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class PluginLoader {
    private static final Map<String, Plugin> MODS = new HashMap<>();
    private static int COUNT = 0;

    public static void onBotStarted() {
        MODS.values().forEach(Plugin::onBotStarted);
    }

    public static void onBotStopped() {
        MODS.values().forEach(Plugin::onBotStopped);
    }

    public static void onPluginUnload() {
        MODS.entrySet().removeIf((entry) -> {
            entry.getValue().onPluginUnload();
            return true;
        });

        MODS.values().forEach(Plugin::onPluginUnload);
    }

    public static void loadPlugins() throws Exception {
        Collection<File> jars = getPluginJars();
        COUNT = jars.size();
        Console.INSTANCE.getLogger().info((COUNT == 0) ? "没有插件被加载" : "正在加载 " + COUNT + " 个插件");
        for (File jar : jars) {
            Plugin plugin = getPlugin(jar);
            MODS.put(plugin.getInfo().getId(), plugin);
            Console.INSTANCE.getLogger().info(String.format("插件 %s %s 已加载", plugin.getInfo().getName(), plugin.getInfo().getVersion()));
        }
    }

    public static int getCount() {
        return COUNT;
    }

    private static Collection<File> getPluginJars() {
        File[] files = ConsoleFiles.PLUGINS.getFile().listFiles();
        if (files == null) return Collections.emptyList();
        return Arrays.stream(files).filter((file) -> file.isFile() && file.getName().endsWith(".jar") && !file.getName().startsWith(".") && !file.isHidden()).collect(Collectors.toList());
    }

    private static Plugin getPlugin(File file) throws Exception {
        URI uri = file.toURI();
        URI jarURI = new URI("jar:" + uri.getScheme(), uri.getHost(), uri.getPath(), uri.getFragment());
        FileSystem fileSystem = FileSystems.newFileSystem(jarURI, Maps.newHashMap());
        Path path = fileSystem.getPath("plugin.json");
        InputStream inputStream = Files.newInputStream(path);
        InputStreamReader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
        JsonObject json = Util.INSTANCE.getGSON().fromJson(reader, JsonObject.class);
        inputStream.close();
        reader.close();
        PluginInformation info = new PluginInformation(json);
        if (MODS.containsKey(info.getId())) throw new PluginException("插件 " + file + " [" + String.format("%s (%s) v%s", info.getName(), info.getId(), info.getVersion()) + "] 无法加载, 已有同名插件 " + MODS.get(info.getId()));
        String main = info.getMain();
        URLClassLoader ucl = new URLClassLoader(new URL[]{new URL("jar:" + uri.toURL() + "!/")});
        Class<?> clazz = Class.forName(main, true, ucl);
        Plugin plugin = (Plugin) clazz.getDeclaredConstructor().newInstance();
        plugin.init(info);
        return plugin;
    }

//    private static void loadClassPathPlugin() throws Exception {
//        if (isDevelopment()) {
//            ClassLoader loader = PluginLoader.class.getClassLoader();
//            InputStream resource = loader.getResourceAsStream("plugin.json");
//            InputStreamReader reader = new InputStreamReader(resource, StandardCharsets.UTF_8);
//            JsonObject json = Util.INSTANCE.getGSON().fromJson(reader, JsonObject.class);
//            resource.close();
//            reader.close();
//            PluginInformation info = new PluginInformation(json);
//            if (MODS.containsKey(info.getId())) throw new PluginException("插件 " + info.getName() + " [" + String.format("%s (%s) v%s", info.getName(), info.getId(), info.getVersion()) + "] 无法加载, 已有同名插件 " + MODS.get(info.getId()));
//            String main = info.getMain();
//            Class<?> clazz = Class.forName(main, true, loader);
//            Plugin plugin = (Plugin) clazz.getDeclaredConstructor().newInstance();
//            plugin.init(info);
//            MODS.put(plugin.getInfo().getId(), plugin);
//            Console.INSTANCE.getLogger().info(String.format("插件 %s %s 已加载", plugin.getInfo().getName(), plugin.getInfo().getVersion()));
//
//        }
//    }
//
//
}

File: src/main/kotlin/net/cjsah/console/plugin/PluginLoader.kt
package net.cjsah.console.plugin

import com.google.gson.JsonObject
import net.cjsah.console.Console
import net.cjsah.console.Console.logger
import net.cjsah.console.ConsoleFiles
import net.cjsah.console.Util
import net.cjsah.console.command.CommandManager
import net.cjsah.console.exceptions.BuiltExceptions.PLUGIN_UNKNOWN_EXCEPTION
import net.cjsah.console.exceptions.PluginException
import java.io.File
import java.io.InputStreamReader
import java.net.URLClassLoader
import java.nio.charset.StandardCharsets
import java.util.jar.JarFile

object PluginLoader {

    fun onBotStarted() = Console.plugins.values.forEach { it.onBotStarted() }

    fun onBotStopped() = Console.plugins.values.forEach {
        CommandManager.deregister(it)
        it.onBotStopped()
    }

    fun onPluginUnload() = Console.plugins.entries.removeAll { (_, plugin) ->
        plugin.onPluginUnload()
        true
    }

    @Throws(Exception::class)
    fun loadPlugins() {
        val jarFiles = getPluginJars()
        logger.info(if (jarFiles.isEmpty()) "没有插件被加载" else "正在加载 ${jarFiles.size} 个插件")
        jarFiles.forEach { file ->
            getPlugin(file).let {
                Console.plugins[it.getInfo().id] = it
                Console.permissions.addPlugin(it)
                logger.info("插件 ${it.getInfo().name} ${it.getInfo().version} 已加载")
            }
        }
    }

    fun getPlugin(id: String): Plugin? = Console.plugins[id]

    private fun getPluginJars(): Collection<File> =
        ConsoleFiles.PLUGINS.file.listFiles().filter { it.isFile && it.extension == "jar" }

    @Throws(Exception::class)
    private fun getPlugin(file: File): Plugin {
        val classLoader = URLClassLoader(arrayOf(file.toURI().toURL()))
        val inputStream = JarFile(file).let { it.getInputStream(it.getJarEntry("plugin.json")) }
        val reader = InputStreamReader(inputStream, StandardCharsets.UTF_8)
        val info = PluginInformation(Util.GSON.fromJson(reader, JsonObject::class.java))
        inputStream.close()
        reader.close()
        if (Console.plugins.containsKey(info.id))
            throw PluginException("插件 $file [${info.name} (${info.id}) v${info.version}] 无法加载, 已有同名插件 ${Console.plugins[info.id]}")
        if (!info.isAvailableVersion())
            throw PluginException("插件 $file [${info.name} (${info.id}) v${info.version}] 无法加载, 插件版本与控制台版本不一致")
        return (classLoader.loadClass(info.main).getDeclaredConstructor().newInstance() as Plugin).also { it.init(info) }
    }

    private fun PluginInformation.isAvailableVersion(): Boolean {
        if (this.depends.containsKey("console")) {
            val consoleVersions = Console.version.split(".").map { it.toInt() }
            val pluginVersion = this.depends["console"]!!
            val pluginPrefix = pluginVersion.substring(0, 1)
            val pluginSuffix = pluginVersion.substring(pluginVersion.length - 1, pluginVersion.length)
            val pluginVersions = pluginVersion.substring(1, pluginVersion.length - 1).split(".").map { it.toInt() }
            val prefixRule: (Int, Int) -> Boolean = when (pluginPrefix) {
                "[" -> { a, b -> a >= b }
                "(" -> { a, b -> a > b }
                else -> throw PLUGIN_UNKNOWN_EXCEPTION.create()
            }
            val suffixRule: (Int, Int) -> Boolean = when (pluginSuffix) {
                "]" -> { a, b -> a <= b }
                ")" -> { a, b -> a < b }
                else -> throw PLUGIN_UNKNOWN_EXCEPTION.create()
            }
            for (index in pluginVersions.indices) {
                if (!(prefixRule(pluginVersions[index], consoleVersions[index]) &&
                      suffixRule(pluginVersions[index], consoleVersions[index])))
                    return false
            }
            return true
        } else return true
    }
}